californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -604,7 +604,12 @@ public void receiveResponse(final Exchange exchange, final Response response) {
 								Level.FINER,
 								"discarding old block2 transfer [{0}], received during ongoing block2 transfer {1}",
 								new Object[]{ response, status.getObserve() });
-						exchange.setComplete();
+						// We don't complete older exchange 
+						// if this is the one used for more recent block2 exchange
+						if(!status.getExchange().equals(exchange))
```

Comment:
Should we do this test for `exchange.setComplete()` below ?

Reply:
I'm not sure, if introducing the usage of `Exchange.equals()`, is needed. 
I miss the implementation of equals in Exchange and until now `==` seems to work.

But I have to see the issue tomorrow more detailed.

Reply:
agreed to  `==` instead of `equals`.

Reply:
final void completeNewTransfer(Exchange newExchange) {
		Exchange exchange;
		synchronized (this) {
			exchange = this.exchange;
		}
		if (exchange != null && newExchange != exchange) {
			// complete new exchange
			newExchange.setComplete();
		}
	}

What's about add a new method instead of expose the exchange?

Reply:
Yes, debugging with your new test shows, that it was a bug to always complete the exchange.
But I would prefer to stick with the pattern of named "complete" in the status rather then export the exchange.

Reply:
Ok, I will do that.

You didn't answer to : "Should we do this test for exchange.setComplete() below ?"

Reply:
> Yes, debugging with your new test shows, that it was a bug to always complete the exchange. 

So, yes, we should add that test.

Reply:
Ok thx